🎨  logging improvements

kirrg001 opened the pull request:
refs #7116

This PR contains small improvements to our logging component.
I already marked the `TODOS` in the issue #7116 !

If you go step by step through the commits, every single change is hopefully clear. Please read the commit description.

My favourite way of developing with node:
`LEVEL=error MODE=long node index.js`
It only prints errors and all request/response information.

kirrg001 commented:
ready for review

ErisDS commented on core/server/app.js:
```diff
@@ -22,12 +23,18 @@ module.exports = function setupParentApp() {
     // (X-Forwarded-Proto header will be checked, if present)
     parentApp.enable('trust proxy');
 
+    parentApp.use(responseTime());
```
Can we please (please, please) use the same request logging middleware we already have in our internal code?

This `responseTime` dependency adds a header, which is very different to the intention of the TODO, which was to fix https://github.com/TryGhost/Ghost/pull/7597/commits/0541af87ad81176d9ba3d77c793a3281cd0db02f#diff-85567d8f2026caaf1a67588fb1641aa9R32 which is IMO too simplistic and to also add duration to the logs.

The middleware I'm suggesting we use:
- adds duration without needing a dependency
- also adds an id for requests, so multiple logs from the same request are traceable
    - note this requires a dependency for generating a uuid, but we can use the same one we will use for creating a UID for errors and could even use the uid util that already exists
- has more advanced handling of request/responses borrowed from morgan, which covers more edge-cases for the behaviour of requests and responses.

kirrg001 replied:
I would like to keep express formatting. See:

### with express:
```
INFO undefined [2016-10-20 12:12:33] "GET /ghost/img/ghosticon.jpg" 301 3.447ms
INFO 127.0.0.1 [2016-10-20 12:12:33] "GET /" 200 154.220ms
INFO 127.0.0.1 [2016-10-20 12:12:33] "GET /assets/css/screen.css?v=8fcd3ab6e5" 200 9.465ms
INFO 127.0.0.1 [2016-10-20 12:12:33] "GET /assets/js/jquery.fitvids.js?v=8fcd3ab6e5" 200 3.092ms
INFO 127.0.0.1 [2016-10-20 12:12:33] "GET /assets/js/index.js?v=8fcd3ab6e5" 200 2.878ms
INFO 127.0.0.1 [2016-10-20 12:12:33] "GET /assets/fonts/casper-icons.woff?v=1" 304 2.789ms
```

### with our internal solution:
```
INFO undefined [2016-10-20 12:13:07] "GET /ghost/img/ghosticon.jpg" 301 3
INFO 127.0.0.1 [2016-10-20 12:13:07] "GET /" 200 163
INFO 127.0.0.1 [2016-10-20 12:13:07] "GET /assets/css/screen.css?v=e389b8ca44" 200 8
INFO 127.0.0.1 [2016-10-20 12:13:07] "GET /assets/js/jquery.fitvids.js?v=e389b8ca44" 200 3
INFO 127.0.0.1 [2016-10-20 12:13:07] "GET /assets/js/index.js?v=e389b8ca44" 200 3
INFO 127.0.0.1 [2016-10-20 12:13:07] "GET /assets/fonts/casper-icons.woff?v=1" 304 2
```

ErisDS replied:
Cool, that should be possible by swapping `Date.now()` for `process.hrtime()`

kirrg001 replied:
It's way more code then just replacing a single line.

```
var startTime = process.hrtime();
var endTime = process.hrtime(startTime);
> [ 0, 11835251 ]
var time = endTime[0] * 1e3 + endTime[1] * 1e-6
> 9.506681
time.toFixed(3) + 'ms'
> 9.50ms
```

I would like to avoid copying their logic over to Ghost. What do you think?

ErisDS commented on core/server/logging/index.js:
```diff
@@ -3,8 +3,8 @@ var config = require('../config'),
-        mode: process.env.NODE_MODE || config.get('logging:mode'),
-        level: process.env.NODE_LEVEL || config.get('logging:level'),
+        mode: process.env.NODE_MODE || process.env.MODE || config.get('logging:mode'),
+        level: process.env.NODE_LEVEL || process.env.LEVEL || config.get('logging:level'),
```
This is cool 👍

ErisDS commented on core/server/auth/passport.js:
```diff
@@ -101,7 +102,7 @@ exports.init = function initPassport(options) {
             url: utils.url.getBaseUrl()
         }).then(function setClient(client) {
-            logging.debug('Public Client Registration was successful');
+            debug('Public Client Registration was successful');
```
Do like 👍

ErisDS commented on core/server/logging/PrettyStream.js:
```diff
@@ -68,84 +68,77 @@ PrettyStream.prototype.write = function write(data) {
-            time,
-            logLevel
-        );
+        if (this.mode !== 'short') {
```
This is a TODO - we should probably move this if statement to prevent the calculation of bodyPretty if it is not going to be output.

ErisDS commented on core/server/logging/PrettyStream.js:
```diff
@@ -68,84 +68,77 @@ PrettyStream.prototype.write = function write(data) {
+        // can be extended to define from outside, but not important
+        try {
+            output += format('%s %s [%s] "%s %s" %s %s\n',
+                logLevel,
```
This looks really awesome 👍 
I'm not entirely sure that having 127.0.0.1 output when developing locally is useful? Maybe something to consider for later, maybe it's only useful to include in the JSON output for production?

Also note there is a :bug: where I am seeing `undefined` for `data.req.ip` when first starting the server and hitting the blog.

kirrg001 replied:
> Also note there is a 🐛 where I am seeing undefined for data.req.ip when first starting the server and hitting the blog.

I have no idea why it's undefined :/

We could delete the ip param for now?don't find it very helpful anyway.

ErisDS replied:
Yeah I don't think its helpful in dev mode.

kirrg001 commented:
https://github.com/TryGhost/Ghost/pull/7597/commits/74a8e2901b99159466ead854c991b23000f5bd77

Removed req.ip from log output.
Removed response-time dependency.

kirrg001 commented:
pushed moving request logging into a middleware

 https://github.com/TryGhost/Ghost/pull/7597/commits/82f9265a1141d1ed753a55d271ba162196403951

kirrg001 commented:
I have added a new commit: https://github.com/TryGhost/Ghost/pull/7597/commits/d7389e4bb13d909f7e19348a96aa6f5f19250f02

See commit description. We add a requestId and userId to each log entry if available.
If there is no userId available, it will show `null` right now.

kirrg001 commented:
Removed `cuid` for requestId as suggested. Using `uuid` now instead 👍 Thanks
